sympy/tensor/array/ndim_array.py

```diff
@@ -143,6 +143,9 @@ def __new__(cls, iterable, shape=None, **kwargs):
         from sympy.tensor.array import ImmutableDenseNDimArray
         return ImmutableDenseNDimArray(iterable, shape, **kwargs)
 
+    def __getitem__(self, index):
+        ...
```

Comment:
What does this do?

Reply:
Wouldn't this mean that the code (either before or after this PR) would fail if an `NDimArray` (and not one of its subclasses) was printed for some reason and has `.rank == 0`?

Reply:
The `__new__` function returns an `ImmutableDenseNDimArray`, so if you call `NDimArray([])` it won't return a base class instance.

Reply:
I was under the impression that baseclasses in sympy usually raise NotImplementedError if their methods are not well defined in the base class (see e.g. `Number` in core.numbers)

Reply:
I don't think `NotImplementedError` is a proper way for this. It should be used for real not-implemented branch of an algorithm. If we do want to prevent a base class instantiation, we should use `abc`:
```py
>>> import abc
>>> class C(abc.ABC):
...     @abc.abstractmethod
...     def foo(self):
...             return 0
... 
>>> C()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class C with abstract method foo
```

Reply:
This still hasn't been adressed

Reply:
Thanks. Converted to abstractmethod.